src/main/java/org/breedinginsight/brapi/v2/BrAPIObservationsController.java

```diff
@@ -93,8 +106,96 @@ public HttpResponse observationsGet(@PathVariable("programId") UUID programId,
+                    observationUnitLevelRelationshipDbId,
+                    commonCropName,
+                    programDbId,
+                    trialDbId,
+                    germplasmDbId,
+                    externalReferenceID,
+                    externalReferenceId,
+                    externalReferenceSource
+            ).filter(Objects::nonNull).collect(Collectors.toList());
+
+            if (!unsupportedParams.isEmpty()) {
+                return HttpResponse.status(HttpStatus.NOT_IMPLEMENTED).body(
+                        new BrAPIObservationListResponse().metadata(new BrAPIMetadata().status(List.of(new BrAPIStatus().messageType(BrAPIStatus.MessageTypeEnum.ERROR)
+                                .message("Unsupported query parameter. Only studyDbId, page, and pageSize are supported."))))
+                );
+            }
+
+            // Get a filtered list of observations.
+            List<BrAPIObservation> observations = observationDAO.getObservationsByFilters(program.get(), studyDbId);
+
+            // Total number of records in the unpaged super set.
+            int totalCount = observations.size();
+            // Zero-indexed page, default to zero.
+            int actualPage = page != null ? page : 0;
+            // The least of pageSize and totalCount, unless pageSize is null or zero, in which case use totalCount.
+            int requestedPageSize = (pageSize != null && pageSize > 0) ? Math.min(pageSize, totalCount) : totalCount;
+            // Integer division and round up.
+            int totalPages = totalCount / requestedPageSize + ((totalCount % requestedPageSize == 0) ? 0 : 1);
+            log.info("(Pagination) totalCount: " + totalCount + " actualPage (0-indexed): " + actualPage + " requestedPageSize: " + requestedPageSize + " totalPages: " + totalPages);
+
+            // Determine validity of pagination query parameters.
+            boolean pageSizeValid = pageSize != null && pageSize > 0;
+            boolean pageValid = page != null && page >= 0 && page < totalPages;
+
+            // Only paginate if valid pagination values were sent.
+            if (pageSizeValid && pageValid) {
+                int start = actualPage * requestedPageSize;
+                // Account for last page, which may have fewer than requestedPageSize items, or exactly requestedPageSize items.
+                int end = (actualPage == (totalPages - 1) && totalCount % requestedPageSize != 0) ? (start + (totalCount % requestedPageSize)) : Math.min(((actualPage + 1) * requestedPageSize), totalCount);
+                log.info("(Pagination) start " + start + " end " + end);
+                // Sort observations so that paging is consistent and coherent.
+                observations.sort(Comparator.comparing(BrAPIObservation::getObservationDbId));
+                // Paginate response.
+                observations = observations.subList(start, end);
+            } else if (pageSize != null || page != null) {
```

Comment:
This behavior differs from the brapi server, is that what we want? For example if you specify pageSize without page it will return a response with pageSize objects and assume page 0.

Reply:
Sounds good